### PR TITLE
Add more lints for conda recipes

### LINKED ISSUE #469: Missing list delimiter in maintainers list breaks things
This is basically a "this one thing I accidentally did broke the system once" bug.

In particular, I accidentally wrote the maintainers list of my recipe like so:

```
extra:
  recipe-maintainers:
    pkgw
```

And then bad things happened. See [here](https://github.com/conda-forge/libsakura-feedstock/pull/1), [here](https://github.com/conda-forge/staged-recipes/pull/2530).

@jakirkham fixed it all up, but the linter should catch this!

### LINKED ISSUE #520: Linter should check the recipe name
Ref: https://github.com/conda-forge/staged-recipes/pull/3004#discussion_r117669184

### PR DESCRIPTION
**Pull request by isuruf**
Fixes #520 
Fixes #469 

There are more lints when env variable, `CONDA_FORGE_LINTS` is set. It checks that the github usernames are valid and that there is no existing feedstock.

### CODE REVIEW

**Comment by jakirkham** [comment]
Nice additions. Could we please try using a CLI flag instead of environment variable?

**Comment by isuruf** [comment]
@jakirkham, can you have another look?

File: conda_smithy/lint_recipe.py [hunk 1]
```diff
@@ -149,9 +154,56 @@ def lintify(meta, recipe_dir=None):
...
+
+    # 15: Run conda-forge specific lints
+    if conda_forge:
+        run_conda_forge_lints(meta, recipe_dir, lints)
+
     return lints
 
 
+def run_conda_forge_lints(meta, recipe_dir, lints):
+    import github
```
Comment (jakirkham) [inline conda_smithy/lint_recipe.py]: We should be able to import this at the top-level as it is a `conda-smithy` requirement.

**Comment by jakirkham** [comment]
How do you propose we test the GitHub related tests? Could we use `mock` to do this?

**Comment by isuruf** [comment]
> How do you propose we test the GitHub related tests? Could we use mock to do this?

No, we can use `support` as a non-existent user. Just need a `GH_TOKEN` in .travis.yml to make github calls.

File: conda_smithy/lint_recipe.py (hunk 1 above)
Comment (isuruf) [inline conda_smithy/lint_recipe.py]: Done

**Comment by isuruf** [comment]
Added tests for conda-forge lints. They are skipped unless GH_TOKEN is set.

**Comment by isuruf** [comment]
ping

File: conda_smithy/lint_recipe.py [hunk 2]
```diff
@@ -76,6 +78,10 @@ def lintify(meta, recipe_dir=None):
         lints.append('The recipe could do with some maintainers listed in '
```
Comment (jakirkham) [inline conda_smithy/lint_recipe.py]: Should rename this to 3a.

File: conda_smithy/lint_recipe.py [hunk 3]
```diff
@@ -149,9 +155,55 @@ def lintify(meta, recipe_dir=None):
     except RuntimeError as e:
         lints.append(str(e))
 
+    # 13: Check that the recipe_dir has the same name as the recipe_name
+    recipe_name = package_section.get('name', '').strip()
+    if recipe_dir:
```
Comment (jakirkham) [inline conda_smithy/lint_recipe.py]: So I don't think we want to enforce this in the feedstocks. As for staged-recipes, it's suppose to not care what the directories are named any more. ( https://github.com/conda-forge/staged-recipes/pull/2313 ) Have there been cases where this is not working?

File: conda_smithy/lint_recipe.py (hunk 2 above)
Comment (jakirkham) [inline conda_smithy/lint_recipe.py]: Done.

**Comment by jakirkham** [comment]
When it comes to testing the GitHub dependent conda-forge features, adding a secure environment variable will suppress it from showing up in a PR. It can be worked around by pushing any PR branch onto a branch on this repo. Though we already do this over at [conda-forge-webservices]( https://github.com/conda-forge/conda-forge-webservices ). Given the scope of these particular changes (limited to conda-forge and require the GitHub API), I'm wondering whether it would be better to have them here or add them to the webservice's linter outright. Thoughts on this?

File: conda_smithy/lint_recipe.py (hunk 3 above)
Comment (isuruf) [inline conda_smithy/lint_recipe.py]: That explains it. I haven't seen this error in a while. I'll remove it. (Btw, for feedstocks, the `recipe_dirname != 'recipe'` is false and this check is not performed. See line 163)

**Comment by ocefpaf** [comment]
@isuruf is this good to go?

**Comment by isuruf** [comment]
This is ready now

File: conda_smithy/lint_recipe.py [hunk 4]
```diff
@@ -71,11 +73,15 @@ def lintify(meta, recipe_dir=None):
             lints.append('The {} item is expected in the about section.'
                          ''.format(about_item))
 
-    # 3: The recipe should have some maintainers.
+    # 3a: The recipe should have some maintainers.
     if not extra_section.get('recipe-maintainers', []):
         lints.append('The recipe could do with some maintainers listed in '
                      'the `extra/recipe-maintainers` section.')
 
+    # 3b: Maintainers should be a list
```
Comment (ocefpaf) [inline conda_smithy/lint_recipe.py]: We should drop this numbering altogether. It only gets in the way when updating/adding more linting rules.

**Comment by ocefpaf** [comment]
@isuruf LGTM I re-started Travis-CI to see if the failing status had something to do with conda version. Merging if we get green :wink:

**Comment by isuruf** [comment]
Yes, the failure is because of conda version. @jakirkham, this requires a pinning or removing `conda=4.3.23`. IMHO, removing `conda=4.3.23` from `conda-forge` is better than pinning it everywhere.

**Comment by isuruf** [comment]
All tests pass now

**Pull request merged**